coral/src/app/features/topics/schema-request/components/TopicSchema.tsx

```diff
@@ -0,0 +1,151 @@
+  const { name, required } = props;
+  const { setValue, setError, clearErrors, control } =
+    useFormContext<TopicRequestFormSchema>();
+
+  const [schema, setSchema] = useState<string | undefined>(undefined);
+
+  useEffect(() => {
+    if (isString(schema)) {
+      setValue(name, schema, {
+        shouldValidate: true,
+        shouldTouch: true,
+        shouldDirty: true,
+      });
+    }
+  }, [schema]);
+
+  function checkEmptyFile(event: ChangeEvent<HTMLInputElement>) {
+    if (!required) return;
+    const file = event.target?.files?.[0];
+    if (!file) {
+      setError(name, { message: "File is a required field" });
+    }
+  }
+
+  function validateSchema(markers: editor.IMarker[]) {
+    if (markers.length > 0) {
+      setError(name, { message: markers?.[0]?.message, type: "custom" });
+      setValue(name, undefined, {
+        shouldValidate: false,
+        shouldDirty: true,
+        shouldTouch: true,
+      });
+    }
+  }
+
+  function uploadFile(event: ChangeEvent<HTMLInputElement>) {
+    const file = event.target?.files?.[0];
+    if (file) {
+      // eslint-disable-next-line no-inner-declarations
+      async function getFile() {
+        return await readFile(file);
+      }
+      getFile().then((fileContent) => {
+        clearErrors();
+        setSchema(fileContent);
+      });
+    } else {
+      setError(name, { message: "File is a required field" });
+      setSchema(undefined);
+    }
+  }
+
+  return (
+    <_Controller
+      name={name}
+      control={control}
+      render={({ fieldState: { error } }) => {
+        return (
+          <div>
+            <FileInput
+              {...props}
+              buttonText={"Upload AVRO schema"}
+              labelText={"Upload AVRO schema file"}
+              noFileText={"No file chosen"}
+              helperText={error?.message || ""}
+              required={required}
+              onBlur={checkEmptyFile}
+              onChange={uploadFile}
+              valid={!error}
+            />
+            <Typography.Caption fontWeight={"500"} htmlTag={"label"}>
+              <span className={error?.message && "text-error-50"}>
+                Schema preview (read only)
+              </span>
+            </Typography.Caption>
+            <BorderBox
+              borderColor={error?.message ? "error-50" : "grey-20"}
+              marginTop={"1"}
+            >
+              {!schema && (
+                <div style={{ height: "200px" }}>
+                  <Box
+                    display={"flex"}
+                    justifyContent={"center"}
+                    alignItems={"center"}
+                    height={"full"}
+                    backgroundColor={"grey-5"}
+                    aria-hidden={"true"}
+                  >
+                    Preview for your schema
+                  </Box>
+                </div>
+              )}
+              <div style={{ display: `${!schema ? "none" : "block"}` }}>
```

Comment:
I would have implemented with with:

```
{ isString(schema) && <Monaco ...}
```

That way we would not need to touch the `style` property, which usually indication that the component library is not flexible for the desired design.

Reply:
Let me check if that works now. I had to do that because if the Monaco editor was only rendered when the value was already set, it did not trigger the onChange which I needed at that point. 

But now that I did the validation differently and don't do anything onChange, it should not be needed anymore. 

(this is why code reviews are so valuable, sometimes you just can't that kind of thing anymore ❤️ )